File: libs/std/intify.lua
require "SltRuntime"

function isNumLua(n) return tonumber(n().value) == nil end

isNumBase = SltThunk.create(
    function() return
        SltFunc.create(
            function(n) return SltBool.create(tonumber(tostring(n())) ~= nil, n().loc) end
        )
    end
)

stringToInt = SltThunk.create(
    function() return
        SltFunc.create(
            function(i)
                if isNumLua(i) then 
                    return SltError.create("TypeError: ", "First argument is not a string", i())
                else
                    return SltNum.create(tonumber(i().value), {6, 1, "intify.lua"})
                end
            end
            ),
            {7, 1, "intify.lua"}
    end
)

File: libs/std/intify.js
import {SltThunk, SltFunc, SltNum, SltBool, error} from "../../SltRuntime.ts"

const isNum = n => !isNaN(n().toString()) && !isNaN(parseFloat(n().toString()))

const isNumBase = new SltThunk(
    () => new SltFunc(n => new SltBool(isNum(n), n().loc), [7, 2, "intify.js"])
)

const stringToInt = new SltThunk(
    () =>
        new SltFunc(
            i => {
                    if (!isNum(i)) error("TypeError: ", "First argument is not a string convertible to number", i())
                    else return new SltNum(parseFloat(i().toString()), [6, 1, "intify.js"])
                }, 
                [2, 5, "intify.js"]
            )
)

export {stringToInt, isNumBase}